File: prism/wp.py
"""Frama-C WP-shaped weakest-precondition discharge.

Mined from third_party/Frama-C/src/plugins/wp:

- calculus.ml `get_weakest_precondition`: the postcondition is a goal
  pushed backward through the CFG; `requires` is a hypothesis, not a
  proof of the body on the whole domain.
- cfgDump.ml: call-pre is "Prove PreCond"; post is folded as hyp.
- wp_error.ml `unsupported`: a construct the model does not encode
  aborts the calculus — that is ERROR, never a closed proof.
- VC.ml: ACSL requires/ensures become verification conditions (VCs).
- QED (wp_behav.c `qed_ok`) discharges trivial VCs without a prover.

This is PRISM's in-tree engine. Do not wrap the `frama-c` binary
(adapters_extra `_run_frama_c` is EVA, and a missing binary is NOTRUN).

A closed check is PROVED-ASSUMING, never PROVED / PROVED-UNBOUNDED:
requires is an explicit harness. POINTER stays NEEDS-HARNESS.
Unencodable ACSL (\\valid, \\forall, \\old, ...) is ERROR, not a proof.
"""

from __future__ import annotations

import re

from prism import laws
from prism.cparse import body_needs_pointer_harness
from prism.contracts import bmc_function_with_assume, parse_comments
from prism.models import Finding, FunctionInfo

_RETURN = re.compile(r"\breturn\s+([^;]+);")

# ACSL logic tokens Frama-C WP encodes in a memory/logic model we do not have.
# `\result` is the one exception: it is the returned scalar.
_ACSL_UNENC = re.compile(
    r"\\(?:"
    r"valid(?:_read|_write|_index)?|forall|exists|old|at|"
    r"separated|initialized|dangling|null|base_addr|block_length|"
    r"offset|allocable|freeable|fresh|from|nothing|let|lambda|"
    r"true|false|union|inter|subset|empty|is_finite|is_NaN|"
    r"min|max|sum|product|numof|matches|unspecified|allocation"
    r")\b",
    re.I,
)

_TOK = re.compile(
    r"\s+|"
    r"==>|==|!=|<=|>=|&&|\|\||<<|>>|->|"
    r"[+\-*/%<>=!&|^~?:()]|"
    r"0[xX][0-9A-Fa-f]+|"
    r"\d+|"
    r"[A-Za-z_]\w*|"
    r"."
)

_IDENT = re.compile(r"^[A-Za-z_]\w*$")
_NUM = re.compile(r"^(?:0[xX][0-9A-Fa-f]+|\d+)$")
_BINOP = {
    "+", "-", "*", "/", "%", "<", ">", "<=", ">=", "==", "!=",
    "&&", "||", "<<", ">>", "&", "|", "^", "?", ":",
}
_UNARY_OK = {"+", "-", "!", "~"}
_BEFORE_UNARY = {None, "(", "?", ":"} | _BINOP


def encode_predicate(text: str) -> str | None:
    """Map an ACSL / `// requires:` atom to a C scalar expression, or None.

    None means unencodable: the caller must ERROR, never PROVED-ASSUMING.
    """
    raw = (text or "").strip()
    if not raw:
        return None
    raw = raw.replace("\\result", "result")
    if _ACSL_UNENC.search(raw) or "\\ " in raw or raw.startswith("\\"):
        return None
    if "<==>" in raw or "^^" in raw:
        return None
    rewritten = _rewrite_implies(raw)
    if rewritten is None:
        return None
    return _encode_scalar(rewritten)


def run_wp(functions: list[FunctionInfo], unwind: int = 8) -> list[Finding]:
    out: list[Finding] = []
    for fn in functions:
        spec = parse_comments(fn)
        reqs: list[str] = spec.get("requires") or []
        ens: list[str] = spec.get("ensures") or []
        if not ens:
            continue
        base = dict(
            stage="wp", file=fn.file, function=fn.name, line=fn.line,
            cls="FUNC-CONTRACT", strength=laws.STRENGTH_PROVES,
        )
        if fn.kind == "POINTER" or body_needs_pointer_harness(fn.body or ""):
            out.append(Finding(
                **base, status=laws.NEEDS_HARNESS,
                message=(
                    "POINTER: WP would invent a buffer; Frama-C WP binary "
                    "is not a proof either"
                ),
                extra={"engine": "prism-wp", "wp": "return-substitution"},
            ))
            continue
        if fn.kind != "SCALAR":
            out.append(Finding(
                **base, status=laws.NEEDS_HARNESS,
                message=f"{fn.kind}: WP scalar subset only; not a proof",
                extra={"engine": "prism-wp", "wp": "return-substitution"},
            ))
            continue

        encoded_req: list[str] = []
        bad: str | None = None
        for r in reqs:
            e = encode_predicate(r)
            if e is None:
                bad = r
                break
            encoded_req.append(e)
        encoded_ens: list[str] = []
        if bad is None:
            for e0 in ens:
                e = encode_predicate(e0)
                if e is None:
                    bad = e0
                    break
                encoded_ens.append(e)
        if bad is not None:
            out.append(Finding(
                **base, status=laws.ERROR,
                message=f"cannot encode WP predicate ({bad})",
                extra={
                    "engine": "prism-wp",
                    "wp": "return-substitution",
                    "wp_unencoded": True,
                    "requires": " && ".join(reqs) if reqs else None,
                    "ensures": " && ".join(ens),
                },
            ))
            continue

        requires = " && ".join(encoded_req) if encoded_req else None
        ensures = " && ".join(encoded_ens)
        returns = [m.strip() for m in _RETURN.findall(fn.body or "")][:8]
        vcs = [_subst_result(ensures, expr) for expr in returns] or [ensures]
        extra = {
            "engine": "prism-wp",
            "wp": "return-substitution",
            "wp_returns": returns,
            "wp_vc": (
                (f"({requires}) ==> " if requires else "")
                + " && ".join(f"({v})" for v in vcs)
            ),
            "requires": requires,
            "ensures": ensures,
        }
        if vcs and all(_is_tautology(v) for v in vcs):
            extra["wp_qed"] = True
            out.append(Finding(
                **base, status=laws.PROVED_ASSUMING,
                message=(
                    f"WP of ensures ({ensures}) holds"
                    + (f" assuming ({requires})" if requires else "")
                    + "; never an unconditional PROVED"
                ),
                extra=extra,
            ))
            continue

        rec = bmc_function_with_assume(fn, unwind, requires, ensures)
        rec.stage = "wp"
        merged = dict(rec.extra or {})
        merged.update(extra)
        rec.extra = merged
        if laws.is_proof(rec.status):
            was = rec.status
            rec.status = laws.PROVED_ASSUMING
            if was == laws.PROVED_UNBOUNDED:
                rec.message = (
                    f"WP of ensures ({ensures}) holds for all unrollings"
                    + (f" assuming ({requires})" if requires else "")
                    + "; never PROVED-UNBOUNDED from WP"
                )
            else:
                rec.message = (
                    f"WP of ensures ({ensures}) holds"
                    + (f" assuming ({requires})" if requires else "")
                    + "; never an unconditional PROVED"
                )
        out.append(rec)
    return out


def _rewrite_implies(text: str) -> str | None:
    """ACSL `==>` → C `!(lhs) || (rhs)`. Nested rewrite; None on junk."""
    text = text.strip()
    if _outer_parens(text):
        inner = _rewrite_implies(text[1:-1])
        return None if inner is None else f"({inner})"
    split = _split_top(text, "==>")
    if split is None:
        return text
    lhs, rhs = split
    left = _rewrite_implies(lhs)
    right = _rewrite_implies(rhs)
    if left is None or right is None:
        return None
    return f"!({left}) || ({right})"


def _split_top(text: str, sep: str) -> tuple[str, str] | None:
    depth = 0
    i = 0
    n = len(sep)
    while i <= len(text) - n:
        ch = text[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
            if depth < 0:
                return None
        elif depth == 0 and text[i:i + n] == sep:
            return text[:i], text[i + n:]
        i += 1
    return None


def _encode_scalar(text: str) -> str | None:
    toks: list[str] = []
    for m in _TOK.finditer(text):
        t = m.group(0)
        if t.isspace() or not t:
            continue
        toks.append(t)
    if not toks:
        return None
    prev: str | None = None
    i = 0
    while i < len(toks):
        t = toks[i]
        if t in {"[", "]", "{", "}", ".", ",", ";", "@", "\\", "#"}:
            return None
        if t == "->":
            return None
        if _IDENT.match(t):
            nxt = toks[i + 1] if i + 1 < len(toks) else None
            if nxt == "(":
                return None
            prev = t
            i += 1
            continue
        if _NUM.match(t):
            prev = t
            i += 1
            continue
        if t in _UNARY_OK or t in {"*", "&"}:
            unary = prev in _BEFORE_UNARY or prev in _UNARY_OK or prev in {"*", "&"}
            if t in {"*", "&"} and unary:
                return None
            if t in _UNARY_OK and unary:
                prev = t
                i += 1
                continue
        if t in _BINOP or t in {"(", ")"}:
            prev = t
            i += 1
            continue
        return None
    return text.strip()


def _subst_result(ensures: str, expr: str) -> str:
    return re.sub(r"\bresult\b", f"({expr})", ensures)


def _is_tautology(pred: str) -> bool:
    """QED-shaped: both sides of == / >= / <= match after stripping parens."""
    p = pred.strip()
    if p in {"1", "true", "True"}:
        return True
    parts = _split_and(p)
    if len(parts) > 1:
        return all(_is_tautology(x) for x in parts)
    for op in ("==", ">=", "<="):
        split = _split_top(p, op)
        if split is None:
            continue
        a, b = split
        if _norm_expr(a) == _norm_expr(b) and _norm_expr(a):
            return True
    return False


def _split_and(text: str) -> list[str]:
    out: list[str] = []
    depth = 0
    last = 0
    i = 0
    while i < len(text) - 1:
        ch = text[i]
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
        elif depth == 0 and text[i:i + 2] == "&&":
            out.append(text[last:i].strip())
            last = i + 2
            i += 2
            continue
        i += 1
    out.append(text[last:].strip())
    return [p for p in out if p]


def _norm_expr(text: str) -> str:
    s = text.strip()
    while _outer_parens(s):
        s = s[1:-1].strip()
    return re.sub(r"\s+", "", s)


def _outer_parens(s: str) -> bool:
    if not (s.startswith("(") and s.endswith(")")):
        return False
    depth = 0
    for i, ch in enumerate(s):
        if ch == "(":
            depth += 1
        elif ch == ")":
            depth -= 1
            if depth == 0:
                return i == len(s) - 1
    return False

File: src/prism/stages/wp.cpp
// Stage wp (the Python engine prism/wp.py): ensures proved by substituting the returned
// expression; unencodable ACSL is ERROR, never PROVED-ASSUMING.
#include "common.hpp"

namespace prism {
namespace fs = std::filesystem;
using namespace stages_detail;

namespace {
// Python engine prism/wp.py encode_predicate / _ACSL_UNENC. Regex is the honesty
// gate: unencodable ACSL is ERROR, never PROVED-ASSUMING.
static Regex wp_acsl_unenc(
    "(?i)\\\\(?:"
    "valid(?:_read|_write|_index)?|forall|exists|old|at|"
    "separated|initialized|dangling|null|base_addr|block_length|"
    "offset|allocable|freeable|fresh|from|nothing|let|lambda|"
    "true|false|union|inter|subset|empty|is_finite|is_NaN|"
    "min|max|sum|product|numof|matches|unspecified|allocation"
    ")\\b");

static Regex wp_tok(
    "\\s+|==>|==|!=|<=|>=|&&|\\|\\||<<|>>|->|"
    "[+\\-*/%<>=!&|^~?:()]|"
    "0[xX][0-9A-Fa-f]+|"
    "\\d+|"
    "[A-Za-z_]\\w*|"
    ".");

static Regex wp_ident_re("^[A-Za-z_]\\w*$");

static Regex wp_num_re("^(?:0[xX][0-9A-Fa-f]+|\\d+)$");

static Regex wp_result_re("\\bresult\\b");

const std::unordered_set<std::string> kWpBinop = {
    "+", "-", "*", "/", "%", "<", ">", "<=", ">=", "==", "!=",
    "&&", "||", "<<", ">>", "&", "|", "^", "?", ":",
};

const std::unordered_set<std::string> kWpUnaryOk = {"+", "-", "!", "~"};

bool wp_outer_parens(std::string_view s) {
    if (s.size() < 2 || s.front() != '(' || s.back() != ')') return false;
    int depth = 0;
    for (std::size_t i = 0; i < s.size(); ++i) {
        if (s[i] == '(') ++depth;
        else if (s[i] == ')') {
            --depth;
            if (depth == 0) return i + 1 == s.size();
        }
    }
    return false;
}

std::optional<std::pair<std::string, std::string>> wp_split_top(std::string_view text, std::string_view sep) {
    int depth = 0;
    const std::size_t n = sep.size();
    if (n == 0 || text.size() < n) return std::nullopt;
    for (std::size_t i = 0; i + n <= text.size(); ++i) {
        char ch = text[i];
        if (ch == '(') ++depth;
        else if (ch == ')') {
            --depth;
            if (depth < 0) return std::nullopt;
        } else if (depth == 0 && text.substr(i, n) == sep)
            return std::make_pair(std::string(text.substr(0, i)), std::string(text.substr(i + n)));
    }
    return std::nullopt;
}

std::optional<std::string> wp_rewrite_implies(std::string text) {
    text = strip(std::move(text));
    if (wp_outer_parens(text)) {
        auto inner = wp_rewrite_implies(text.substr(1, text.size() - 2));
        if (!inner) return std::nullopt;
        return "(" + *inner + ")";
    }
    auto split = wp_split_top(text, "==>");
    if (!split) return text;
    auto left = wp_rewrite_implies(split->first);
    auto right = wp_rewrite_implies(split->second);
    if (!left || !right) return std::nullopt;
    return "!(" + *left + ") || (" + *right + ")";
}

bool wp_before_unary(const std::optional<std::string>& prev) {
    if (!prev) return true;
    if (*prev == "(" || *prev == "?" || *prev == ":") return true;
    if (kWpBinop.contains(*prev)) return true;
    return false;
}

std::optional<std::string> wp_encode_scalar(const std::string& text) {
    std::vector<std::string> toks;
    for (auto& m : wp_tok.finditer(text)) {
        auto t = m.text;
        if (t.empty()) continue;
        bool sp = true;
        for (char c : t)
            if (!std::isspace(static_cast<unsigned char>(c))) {
                sp = false;
                break;
            }
        if (sp) continue;
        toks.push_back(std::move(t));
    }
    if (toks.empty()) return std::nullopt;
    std::optional<std::string> prev;
    for (std::size_t i = 0; i < toks.size(); ++i) {
        const auto& t = toks[i];
        if (t == "[" || t == "]" || t == "{" || t == "}" || t == "." || t == "," || t == ";" ||
            t == "@" || t == "\\" || t == "#")
            return std::nullopt;
        if (t == "->") return std::nullopt;
        if (fullmatch(wp_ident_re, t)) {
            const std::string* nxt = (i + 1 < toks.size()) ? &toks[i + 1] : nullptr;
            if (nxt && *nxt == "(") return std::nullopt;
            prev = t;
            continue;
        }
        if (fullmatch(wp_num_re, t)) {
            prev = t;
            continue;
        }
        if (kWpUnaryOk.contains(t) || t == "*" || t == "&") {
            bool unary = wp_before_unary(prev) || (prev && (kWpUnaryOk.contains(*prev) || *prev == "*" || *prev == "&"));
            if ((t == "*" || t == "&") && unary) return std::nullopt;
            if (kWpUnaryOk.contains(t) && unary) {
                prev = t;
                continue;
            }
        }
        if (kWpBinop.contains(t) || t == "(" || t == ")") {
            prev = t;
            continue;
        }
        return std::nullopt;
    }
    return strip(text);
}

std::optional<std::string> wp_encode_predicate(std::string raw) {
    raw = strip(std::move(raw));
    if (raw.empty()) return std::nullopt;
    {
        const std::string from = "\\result";
        std::size_t pos = 0;
        while ((pos = raw.find(from, pos)) != std::string::npos) {
            raw.replace(pos, from.size(), "result");
            pos += 6;
        }
    }
    if (wp_acsl_unenc.search(raw) || raw.find("\\ ") != std::string::npos || raw.starts_with("\\"))
        return std::nullopt;
    if (raw.find("<==>") != std::string::npos || raw.find("^^") != std::string::npos) return std::nullopt;
    auto rewritten = wp_rewrite_implies(raw);
    if (!rewritten) return std::nullopt;
    return wp_encode_scalar(*rewritten);
}

std::string wp_subst_result(const std::string& ensures, const std::string& expr) {
    std::string out;
    std::size_t i = 0;
    const std::string repl = "(" + expr + ")";
    for (auto& m : wp_result_re.finditer(ensures)) {
        if (m.spans.empty()) continue;
        auto a = static_cast<std::size_t>(std::max(0, m.spans[0].first));
        auto b = static_cast<std::size_t>(std::max(0, m.spans[0].second));
        if (a < i) continue;
        out.append(ensures, i, a - i);
        out += repl;
        i = b;
    }
    out.append(ensures, i, std::string::npos);
    return out;
}

std::string wp_norm_expr(std::string s) {
    s = strip(std::move(s));
    while (wp_outer_parens(s)) s = strip(s.substr(1, s.size() - 2));
    std::string o;
    for (char c : s)
        if (!std::isspace(static_cast<unsigned char>(c))) o.push_back(c);
    return o;
}

std::vector<std::string> wp_split_and(const std::string& text) {
    std::vector<std::string> out;
    int depth = 0;
    std::size_t last = 0;
    std::size_t i = 0;
    while (i + 1 < text.size()) {
        char ch = text[i];
        if (ch == '(') ++depth;
        else if (ch == ')') --depth;
        else if (depth == 0 && text[i] == '&' && text[i + 1] == '&') {
            auto p = strip(text.substr(last, i - last));
            if (!p.empty()) out.push_back(std::move(p));
            last = i + 2;
            i += 2;
            continue;
        }
        ++i;
    }
    auto tail = strip(text.substr(last));
    if (!tail.empty()) out.push_back(std::move(tail));
    return out;
}

bool wp_is_tautology(const std::string& pred) {
    auto p = strip(pred);
    if (p == "1" || p == "true" || p == "True") return true;
    auto parts = wp_split_and(p);
    if (parts.size() > 1) {
        for (auto& x : parts)
            if (!wp_is_tautology(x)) return false;
        return true;
    }
    for (const char* op : {"==", ">=", "<="}) {
        auto split = wp_split_top(p, op);
        if (!split) continue;
        auto a = wp_norm_expr(split->first);
        auto b = wp_norm_expr(split->second);
        if (!a.empty() && a == b) return true;
    }
    return false;
}

}  // namespace

std::vector<Finding> run_wp(const std::vector<FunctionInfo>& functions, int unwind) {
    std::vector<Finding> out;
    static Regex ret_re("\\breturn\\s+([^;]+);");
    for (auto& fn : functions) {
        auto spec = parse_comments(fn);
        if (spec.ensures.empty()) continue;
        auto mark = [](Finding& f) {
            f.extra["engine"] = "prism-wp";
            f.extra["wp"] = "return-substitution";
        };
        if (fn.kind == "POINTER" || body_needs_pointer_harness(fn.body)) {
            auto f = make_find("wp", laws::NEEDS_HARNESS, fn, "FUNC-CONTRACT",
                               "POINTER: WP would invent a buffer; Frama-C WP binary is not a proof either",
                               laws::STRENGTH_PROVES);
            mark(f);
            out.push_back(std::move(f));
            continue;
        }
        if (fn.kind != "SCALAR") {
            auto f = make_find("wp", laws::NEEDS_HARNESS, fn, "FUNC-CONTRACT",
                               fn.kind + ": WP scalar subset only; not a proof", laws::STRENGTH_PROVES);
            mark(f);
            out.push_back(std::move(f));
            continue;
        }

        std::vector<std::string> encoded_req;
        std::optional<std::string> bad;
        for (auto& r : spec.requires_) {
            auto e = wp_encode_predicate(r);
            if (!e) {
                bad = r;
                break;
            }
            encoded_req.push_back(*e);
        }
        std::vector<std::string> encoded_ens;
        if (!bad) {
            for (auto& e0 : spec.ensures) {
                auto e = wp_encode_predicate(e0);
                if (!e) {
                    bad = e0;
                    break;
                }
                encoded_ens.push_back(*e);
            }
        }
        if (bad) {
            auto f = make_find("wp", laws::ERROR, fn, "FUNC-CONTRACT",
                               "cannot encode WP predicate (" + *bad + ")", laws::STRENGTH_PROVES);
            mark(f);
            f.extra["wp_unencoded"] = "true";
            if (!spec.requires_.empty()) f.extra["requires"] = join_sv(spec.requires_, " && ");
            f.extra["ensures"] = join_sv(spec.ensures, " && ");
            out.push_back(std::move(f));
            continue;
        }

        std::optional<std::string> requires_;
        if (!encoded_req.empty()) requires_ = join_sv(encoded_req, " && ");
        std::string ensures = join_sv(encoded_ens, " && ");
        std::vector<std::string> returns;
        nlohmann::json rets = nlohmann::json::array();
        for (auto& m : ret_re.finditer(fn.body)) {
            auto g = strip(m.group(1));
            if (g.empty()) continue;
            returns.push_back(g);
            rets.push_back(g);
            if (returns.size() >= 8) break;
        }
        std::vector<std::string> vcs;
        if (returns.empty()) vcs.push_back(ensures);
        else
            for (auto& expr : returns) vcs.push_back(wp_subst_result(ensures, expr));
        std::string wp_vc;
        if (requires_) wp_vc += "(" + *requires_ + ") ==> ";
        for (std::size_t i = 0; i < vcs.size(); ++i) {
            if (i) wp_vc += " && ";
            wp_vc += "(" + vcs[i] + ")";
        }
        auto fill_extra = [&](Finding& rec) {
            mark(rec);
            rec.extra["wp_returns"] = rets.dump();
            rec.extra["wp_vc"] = wp_vc;
            rec.extra["requires"] = requires_ ? *requires_ : "";
            rec.extra["ensures"] = ensures;
        };

        bool qed = !vcs.empty();
        for (auto& v : vcs)
            if (!wp_is_tautology(v)) qed = false;
        if (qed) {
            auto f = make_find("wp", laws::PROVED_ASSUMING, fn, "FUNC-CONTRACT",
                               "WP of ensures (" + ensures + ") holds" +
                                   (requires_ ? " assuming (" + *requires_ + ")" : "") +
                                   "; never an unconditional PROVED",
                               laws::STRENGTH_PROVES);
            fill_extra(f);
            f.extra["wp_qed"] = "true";
            out.push_back(std::move(f));
            continue;
        }

        std::optional<std::string> ens_opt = ensures;
        auto rec = bmc_with_assume(fn, unwind, requires_, ens_opt, std::nullopt, std::nullopt);
        rec.stage = "wp";
        fill_extra(rec);
        if (rec.status == laws::PROVED) {
            rec.status = std::string(laws::PROVED_ASSUMING);
            rec.message = "WP of ensures (" + ensures + ") holds" +
                          (requires_ ? " assuming (" + *requires_ + ")" : "") +
                          "; never an unconditional PROVED";
        } else if (rec.status == laws::PROVED_UNBOUNDED) {
            rec.status = std::string(laws::PROVED_ASSUMING);
            rec.message = "WP of ensures (" + ensures + ") holds for all unrollings" +
                          (requires_ ? " assuming (" + *requires_ + ")" : "") +
                          "; never PROVED-UNBOUNDED from WP";
        } else if (rec.status == laws::PROVED_ASSUMING) {
            auto low = lower_copy(rec.message);
            if (low.find("never") == std::string::npos) {
                rec.message = "WP of ensures (" + ensures + ") holds" +
                              (requires_ ? " assuming (" + *requires_ + ")" : "") +
                              "; never an unconditional PROVED";
            }
        }
        out.push_back(std::move(rec));
    }
    return out;
}

}  // namespace prism
